File: 86. Partition List/86. Partition List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
ListNode* pullNextNode(ListNode* node) {
    if (node->next != NULL) {
        ListNode* nextNode = node->next;
        node->next = node->next->next;
        nextNode->next = NULL;
        return nextNode;
    } else {
        return NULL;
    }
}


class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        if (head == NULL) {
            return NULL;
        }
        
        ListNode* greaterNodesHead = new ListNode(0);
        ListNode* greaterNodesTail = greaterNodesHead;
        
        ListNode* listHead = new ListNode(0);
        listHead->next = head;
        ListNode* p = listHead;
        
        while (p->next != NULL) {
            if (p->next->val >= x) {
                ListNode* t = pullNextNode(p);
                greaterNodesTail->next = t;
                greaterNodesTail = greaterNodesTail->next;
            } else {
                p = p->next;
            }
        }
        
        p->next = greaterNodesHead->next;
        
        ListNode* t = listHead->next;
        
        delete greaterNodesHead;
        delete listHead;
        return t;
    }
};

File: 86. Partition List/86. Partition List.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }

use std::cmp::Ordering;
impl Solution {
    pub fn partition(head: Option<Box<ListNode>>, x: i32) -> Option<Box<ListNode>> {
        let mut fake_head = Some(Box::new(ListNode::new(0)));
        let mut p = &mut fake_head;
        let mut fake_head_2 = Some(Box::new(ListNode::new(0)));
        let mut p_2 = &mut fake_head_2;
        let mut head = head;
        while head.is_some() {
            let tail = head.as_mut().unwrap().next.take();
            let val = head.as_ref().unwrap().val;
            if val.cmp(&x) == Ordering::Less{
                p.as_mut().unwrap().next = head;
                p = &mut p.as_mut().unwrap().next;
            } else {
                p_2.as_mut().unwrap().next = head;
                p_2 = &mut p_2.as_mut().unwrap().next;
            }
            head = tail;
        }
        p.as_mut().unwrap().next = fake_head_2.as_mut().unwrap().next.take();
        return fake_head.as_mut().unwrap().next.take();
    }
}
